File: wrappers/elixir/lib/turbotoken/encoding.ex
defmodule TurboToken.Encoding do
  @moduledoc """
  An initialized BPE encoding with loaded rank data.
  """

  alias TurboToken.{Nif, Chat, Registry}

  @type t :: %__MODULE__{
          name: String.t(),
          spec: Registry.encoding_spec(),
          rank_payload: binary()
        }

  defstruct [:name, :spec, :rank_payload]

  @doc """
  Encode text into a list of BPE token IDs.
  """
  @spec encode(t(), String.t()) :: {:ok, [non_neg_integer()]} | {:error, term()}
  def encode(%__MODULE__{rank_payload: rp}, text) when is_binary(text) do
    Nif.encode_bpe(rp, text)
  end

  @doc """
  Decode a list of BPE token IDs back to a UTF-8 binary.
  """
  @spec decode(t(), [non_neg_integer()]) :: {:ok, binary()} | {:error, term()}
  def decode(%__MODULE__{rank_payload: rp}, tokens) when is_list(tokens) do
    Nif.decode_bpe(rp, tokens)
  end

  @doc """
  Count the number of BPE tokens in text without materializing the token list.
  """
  @spec count(t(), String.t()) :: {:ok, non_neg_integer()} | {:error, term()}
  def count(%__MODULE__{rank_payload: rp}, text) when is_binary(text) do
    Nif.count_bpe(rp, text)
  end

  @doc """
  Alias for `count/2`.
  """
  @spec count_tokens(t(), String.t()) :: {:ok, non_neg_integer()} | {:error, term()}
  def count_tokens(enc, text), do: count(enc, text)

  @doc """
  Check if text is within a token limit.
  Returns `{:ok, count}` if within limit, `{:ok, false}` if exceeded.
  """
  @spec is_within_token_limit(t(), String.t(), non_neg_integer()) ::
          {:ok, non_neg_integer() | false} | {:error, term()}
  def is_within_token_limit(%__MODULE__{rank_payload: rp}, text, limit)
      when is_binary(text) and is_integer(limit) and limit >= 0 do
    Nif.is_within_token_limit(rp, text, limit)
  end

  @doc """
  Encode a list of chat messages into token IDs.
  """
  @spec encode_chat(t(), [Chat.chat_message()], keyword()) ::
          {:ok, [non_neg_integer()]} | {:error, term()}
  def encode_chat(%__MODULE__{} = enc, messages, opts \\ []) do
    Chat.encode_chat(enc, messages, opts)
  end

  @doc """
  Count tokens in a list of chat messages.
  """
  @spec count_chat(t(), [Chat.chat_message()], keyword()) ::
          {:ok, non_neg_integer()} | {:error, term()}
  def count_chat(%__MODULE__{} = enc, messages, opts \\ []) do
    Chat.count_chat(enc, messages, opts)
  end

  @doc """
  Check if chat messages are within a token limit.
  """
  @spec is_chat_within_token_limit(t(), [Chat.chat_message()], non_neg_integer(), keyword()) ::
          {:ok, non_neg_integer() | false} | {:error, term()}
  def is_chat_within_token_limit(%__MODULE__{} = enc, messages, limit, opts \\ []) do
    case count_chat(enc, messages, opts) do
      {:ok, count} -> {:ok, if(count <= limit, do: count, else: false)}
      err -> err
    end
  end

  @doc """
  Encode a file's contents into token IDs.
  """
  @spec encode_file_path(t(), String.t()) :: {:ok, [non_neg_integer()]} | {:error, term()}
  def encode_file_path(%__MODULE__{} = enc, path) when is_binary(path) do
    case File.read(path) do
      {:ok, content} -> encode(enc, content)
      {:error, reason} -> {:error, {:file_read, reason}}
    end
  end

  @doc """
  Count tokens in a file using the native file counting path.
  """
  @spec count_file_path(t(), String.t()) :: {:ok, non_neg_integer()} | {:error, term()}
  def count_file_path(%__MODULE__{rank_payload: rp}, path) when is_binary(path) do
    Nif.count_bpe_file(rp, path)
  end
end

File: wrappers/lua/src/turbotoken/encoding.lua
local bridge = require("turbotoken.ffi_bridge")
local chat = require("turbotoken.chat")

local Encoding = {}
Encoding.__index = Encoding

--- Create a new Encoding instance.
-- @param name string Encoding name
-- @param rank_payload string Raw rank file bytes
-- @param spec table EncodingSpec from registry
-- @return Encoding
function Encoding.new(name, rank_payload, spec)
    local self = setmetatable({}, Encoding)
    self._name = name
    self._rank_payload = rank_payload
    self._rank_len = #rank_payload
    self._spec = spec
    return self
end

--- Get encoding name.
function Encoding:name()
    return self._name
end

--- Get vocabulary size.
function Encoding:n_vocab()
    return self._spec.n_vocab
end

--- Get end-of-text token ID.
function Encoding:eot_token()
    return self._spec.special_tokens["<|endoftext|>"]
end

--- Encode text to token IDs.
-- @param text string UTF-8 input text
-- @return table Array of token IDs
function Encoding:encode(text)
    local tokens, _ = bridge.encode_bpe_from_ranks(
        self._rank_payload, self._rank_len,
        text, #text
    )
    return tokens
end

--- Decode token IDs back to text.
-- @param tokens table Array of token IDs
-- @return string UTF-8 text
function Encoding:decode(tokens)
    return bridge.decode_bpe_from_ranks(
        self._rank_payload, self._rank_len,
        tokens, #tokens
    )
end

--- Count tokens in text without materializing the token array.
-- @param text string UTF-8 input text
-- @return number Token count
function Encoding:count(text)
    return bridge.count_bpe_from_ranks(
        self._rank_payload, self._rank_len,
        text, #text
    )
end

--- Alias for count().
function Encoding:count_tokens(text)
    return self:count(text)
end

--- Check if text is within a token limit.
-- @param text string UTF-8 input text
-- @param limit number Maximum token count
-- @return number|nil Token count if within limit, nil if exceeded
function Encoding:is_within_token_limit(text, limit)
    return bridge.is_within_token_limit_bpe_from_ranks(
        self._rank_payload, self._rank_len,
        text, #text,
        limit
    )
end

--- Encode chat messages to token IDs.
-- @param messages table[] Array of ChatMessage tables
-- @param opts table|nil Options
-- @return table Array of token IDs
function Encoding:encode_chat(messages, opts)
    return chat.encode_chat(self, messages, opts)
end

--- Count tokens for chat messages.
-- @param messages table[] Array of ChatMessage tables
-- @param opts table|nil Options
-- @return number Token count
function Encoding:count_chat(messages, opts)
    return chat.count_chat(self, messages, opts)
end

--- Check if chat messages are within a token limit.
-- @param messages table[] Array of ChatMessage tables
-- @param limit number Token limit
-- @param opts table|nil Options
-- @return number|nil Token count if within limit, nil if exceeded
function Encoding:is_chat_within_token_limit(messages, limit, opts)
    return chat.is_chat_within_token_limit(self, messages, limit, opts)
end

--- Encode a file's contents to token IDs.
-- @param file_path string Path to file
-- @return table Array of token IDs
function Encoding:encode_file_path(file_path)
    local tokens, _ = bridge.encode_bpe_file_from_ranks(
        self._rank_payload, self._rank_len,
        file_path, #file_path
    )
    return tokens
end

--- Count tokens in a file without materializing the token array.
-- @param file_path string Path to file
-- @return number Token count
function Encoding:count_file_path(file_path)
    return bridge.count_bpe_file_from_ranks(
        self._rank_payload, self._rank_len,
        file_path, #file_path
    )
end

--- Check if a file's content is within a token limit.
-- @param file_path string Path to file
-- @param limit number Maximum token count
-- @return number|nil Token count if within limit, nil if exceeded
function Encoding:is_file_path_within_token_limit(file_path, limit)
    return bridge.is_within_token_limit_bpe_file_from_ranks(
        self._rank_payload, self._rank_len,
        file_path, #file_path,
        limit
    )
end

return Encoding
